src/bounds.jl

```diff
@@ -115,134 +197,136 @@ function bounds_propagation(m::PODNonlinearModel)
+    can help speed up the subsequent solve in subsequent iterations.
 
-    # TODO this sequence need to be changed
+    Only used in presolve bound tightening
+"""
+function resolve_var_bounds(m::PODNonlinearModel, d::Dict; kwargs...)
 
     # Added sequential bound resolving process base on DFS process, which ensures all bounds are secured.
     # Increased complexity from linear to square but a reasonable amount
     # Potentially, additional mapping can be applied to reduce the complexity
-    for i in 1:length(nonlinear_terms)
-        for bi in keys(nonlinear_terms)
-            if (nonlinear_terms[bi][:id] == i) && (nonlinear_terms[bi][:nonlinear_type] in [:bilinear, :monomial, :multilinear])
-                lifted_idx = nonlinear_terms[bi][:lifted_var_ref].args[2]
-                cnt = 0
-                bound = []
-                for var in bi
-                    cnt += 1
-                    var_idx = var.args[2]
-                    var_bounds = [discretization[var_idx][1], discretization[var_idx][end]]
-                    if cnt == 1
-                        bound = copy(var_bounds)
-                    elseif cnt == 2
-                        bound = bound * var_bounds'
-                    else
-                        bound = diag(bound) * var_bounds'
-                    end
-                end
-                if minimum(bound) > discretization[lifted_idx][1]
-                    discretization[lifted_idx][1] = minimum(bound)
-                end
-                if maximum(bound) < discretization[lifted_idx][end]
-                    discretization[lifted_idx][end] = maximum(bound)
-                end
+    for i in 1:length(m.term_seq)
+        k = m.term_seq[i]
+        if haskey(m.nonconvex_terms, k)
+            nlk = k
+            if m.nonconvex_terms[nlk][:nonlinear_type] in POD_C_MONOMIAL
+                d = basic_monomial_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in [:BININT]
+                d = basic_binint_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in [:BINPROD]
+                d = basic_binprod_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in POD_C_TRIGONOMETRIC
+                d = basic_sincos_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in [:BINLIN]
+                d = basic_binlin_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in [:INTLIN]
+                d = basic_intlin_bounds(m, nlk, d)
+            elseif m.nonconvex_terms[nlk][:nonlinear_type] in [:INTPROD]
+                d = basic_intprod_bounds(m, nlk, d)
+            else
+                error("EXPECTED ERROR : NEED IMPLEMENTATION")
             end
+        elseif haskey(m.linear_terms, k)
+            d = basic_linear_bounds(m, k, d)
+        else
+            error("[RARE] Found homeless term key $(k) during bound resolution.")
```

Comment:
What is a homeless term? Is this a term without a relaxation scheme built into POD

Reply:
So, we use the term keys to resolve lifted variable bounds ( just in case these lifted variables are partitioned by a recursively lifted term). Here is an error scheme for not missing any type of terms. This will be improved with better complexity in the next PR.